### PR TITLE
Cleanup HybridDictionary tests

### PR DESCRIPTION
**Pull request by hughbe**
- Take advantage of xunit
- Shorten tests
- Remove duplicate tests

Coverage remains at 100%
/cc @ianhays

### CODE REVIEW

**Comment by hughbe** [comment]
Test  Innerloop OSX Debug Build and Test

File: src/System.Collections.Specialized/tests/HybridDictionary/ContainsObjTests.cs [hunk 1]
```diff
@@ -3,425 +3,67 @@
...
-            len = valuesShort.Length;
-            hd.Clear();
-            for (int i = 0; i < len; i++)
-            {
-                hd.Add(keysShort[i], valuesShort[i]);
-            }
-
-            Assert.Throws<ArgumentNullException>(() => { hd.Contains(null); });
-
-            // [] Contains() for case-insensitive comparer dictionary
```
Comment (ianhays) [inline src/System.Collections.Specialized/tests/HybridDictionary/ContainsObjTests.cs]: I didn't see any case-sensitivity checks in the new code. Did I miss them again?

File: src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs [hunk 2]
```diff
@@ -3,206 +3,66 @@
...
-            {
-                Assert.False(true, string.Format("Error, Values.Count returned {0} instead of {1}", hd.Values.Count, BIG_LENGTH));
-            }
+            Assert.False(hybridDictionary.IsFixedSize);
+            Assert.False(hybridDictionary.IsReadOnly);
+            Assert.False(hybridDictionary.IsSynchronized);
 
-            hd.Remove(lbls[0]);
-            if (hd.Count != BIG_LENGTH - 1)
+            if (caseInsensitive)
```
Comment (ianhays) [inline src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs]: Here's case sensitivity tests for the getter but I didn't see them for Contains.

**Comment by ianhays** [comment]
Looks good, Hugh! Any particular reason why you wrote tests for API like syncroot, issynhcronized, and other interface API instead of just inheriting the tests [common to Collections?](https://github.com/dotnet/corefx/tree/master/src/Common/tests/System/Collections) I admit they're a bit more complicated than I'd have liked, but they'd save you from rewriting a good number of tests that are common across all IDictionary implementations.

File: src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs]: will readd

**Comment by hughbe** [comment]
Thanks for the review - I didn't think of that - I'll take a look

**Comment by hughbe** [comment]
However, just one comment - surely we want to test these methods for 0, < 10 and > 10 count HydridDictionaries - do these common tests allow this?

**Comment by ianhays** [comment]
>However, just one comment - surely we want to test these methods for 0, < 10 and > 10 count HydridDictionaries - do these common tests allow this?

Yup! All common tests are run on three sizes: 0, 1, and 75. They use a common TestBase that defines a method to use as MemberData for all collections tests. [Link here](https://github.com/dotnet/corefx/blob/master/src/Common/tests/System/Collections/TestBase.NonGeneric.cs)

**Comment by hughbe** [comment]
Thanks Ian. 
Nice one - I really like the plug-n-play nature of the TestBase.

I've had to modify the tests to add some more customization points, and I'm close to getting a clean build.

I've encountered one final hurdle:
untitled

It seems there is a rather niche behavioural difference between the implementation for SortedDictionary and Hashtable in terms of the test case 
`ICollection_NonGeneric_CopyTo_NonZeroLowerBound`

When count = 1, an IndexOutOfRangeException is thrown, but when count = 75, an ArgumentException is thrown.

I reckon this could be considered as a bug - to fix it would require adding a check in CopyTo in HybridDictionary for a non-zero lower bound.

Let me know what you think

**Comment by hughbe** [comment]
Also, do any of these test bases test CopyTo working correctly? I can only find tests for CopyTo_Invalid

File: src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs (hunk 2 above)
Comment (hughbe) [inline src/System.Collections.Specialized/tests/HybridDictionary/CtorTests.cs]: https://github.com/hughbe/corefx/blob/specialized-xunit/src/System.Collections.Specialized/tests/HybridDictionary/AddObjObjTests.cs#L36 I found em!

**Comment by ianhays** [comment]
>I reckon this could be considered as a bug - to fix it would require adding a check in CopyTo in HybridDictionary for a non-zero lower bound.

Yeah, looks like a bug to me. I would add the fix in ListDictionary though since it's supposed to be throwing an ArgumentException there [according to the docs](https://msdn.microsoft.com/en-us/library/system.collections.specialized.listdictionary.copyto(v=vs.110).aspx).

> Also, do any of these test bases test CopyTo working correctly? I can only find tests for CopyTo_Invalid

There are some valid tests at the end of the ICollection.NonGeneric.Tests.cs file: ```ICollection_NonGeneric_CopyTo_ArrayIsLargerThanCollection``` and ```ICollection_NonGeneric_CopyTo_ExactlyEnoughSpaceInArray```

**Comment by ianhays** [comment]
Also definitely feel free to add more tests to these files if you feel they're missing anything. In doing so you'll also be adding tests for every class in System.Collections so you'll probably want to test that library as well :)

**Comment by hughbe** [comment]
Should I fix that as part of this PR or seperately?

Do you think I should delete the existing tests for stuff like GetEnumerator, CopyTo, Contains, Clear etc. in HybridDictionary when I move the files?

**Comment by ianhays** [comment]
>Should I fix that as part of this PR or seperately?

It should probably be in its own PR. 

>Do you think I should delete the existing tests for stuff like GetEnumerator, CopyTo, Contains, Clear etc. in HybridDictionary when I move the files?

If they're not any different (i.e. provide no extra test value) than the common ones then you may as well, yeah.

**Comment by hughbe** [comment]
Cheers, I filed #8065 to fix the inconsistency - is it OK if we hold off on merging this until that is merged to simplify things?

**Comment by ianhays** [comment]
>is it OK if we hold off on merging this until that is merged to simplify things?

Of course! Merge them in whichever order you'd prefer.

**Comment by hughbe** [comment]
I reckon that the PR fixing the IndexOutOfRangeException may take a while to merged, so I just pushed a commit moving some stuff to use the common IDictionaryHelpers

I had to add several new customization points to keep the tests working, but these only affect HybridDictionary.

We should probably merge this first :)

**Comment by ianhays** [comment]
Good call, I think Immo is on vacation right now. 

This LGTM. Thanks!

**Comment by hughbe** [comment]
Test Innerloop OSX Debug Build and Test

**Pull request merged**